Payment approve api

spongeboom opened the pull request:
[[5] 결제 승인 API 개발](https://github.com/f-lab-edu/fpay/issues/5)
- 결제 준비 완료 이후 결제 승인 요청 시 처리 가능하도록 개발

IMQA-SUNGYOUN committed:
결제 승인(approvePayment) API 개발
- 승인 API 개발

IMQA-SUNGYOUN committed:
결제 승인(approvePayment) API
- PaymentTransactionDTO toEntity로 객체 생성하도록 변경
- PaymentTransactionResDTO builder 대신 생성자로 넘겨받도록 변경

IMQA-SUNGYOUN committed:
결제 승인(approvePayment) API
- API 테스트 코드 작성 및 Spring Rest Docs 수정

IMQA-SUNGYOUN committed:
결제 승인(approvePayment) API
- baseTimeEntity를 사용하여 생성일,수정일 자동 수정 되도록 변경

IMQA-SUNGYOUN committed:
결제 승인(approvePayment) API
- 중복 request 방지
  - reddison 사용하여 분산 락을 구현

IMQA-SUNGYOUN committed:
결제 승인(approvePayment) API
- 중복 request 방지
  - redisConfig 설정 추가

spongeboom commented on fpay-api/src/main/java/com/flab/fpay/api/pay/service/PaymentTransactionService.java:
```diff
@@ -0,0 +1,56 @@
+                    !paymentTransactionDTO.getCompanyUserId().equals(paymentRequest.getCompanyUserId()) &&
+                    !paymentTransactionDTO.getProductPrice().equals(paymentRequest.getPaymentPrice())){
+                throw new RuntimeException("결제 요청 정보와 승인 정보가 다릅니다.");
+            }
```
[결제 승인 Service] 
- 승인 전 해당 결제건에 대한 결제 요청 정보를 가져와 같은 결제건인지 비교 ( Custom Error를 따로 두어야할까요? )

spongeboom commented on fpay-api/src/main/java/com/flab/fpay/api/pay/service/PaymentTransactionService.java:
```diff
@@ -0,0 +1,56 @@
+                paymentTransaction = approveMoney.approve(paymentTransactionDTO, paymentRequest);
+            }
+
+            return new PaymentTransactionResDTO(paymentTransaction, paymentRequest);
```
[결제 승인 Service] 
- 결제 요청 정보와 비교 후 결제 Type이 Money일 경우 Money 결제 승인 진행

spongeboom commented on fpay-api/src/main/java/com/flab/fpay/api/pay/approve/ApproveMoney.java:
```diff
@@ -0,0 +1,40 @@
+        payInfoService.savePayInfo(payInfo);
+
+        return paymentTransactionDTO.toEntity(paymentRequest, paymentTypeInfo);
+    }
```
[Money 결제 승인] 
- payment Type이 결제인 경우 결제 정보들을 넘겨 받아 결제 승인 진행

spongeboom commented on fpay-api/src/main/java/com/flab/fpay/api/pay/service/PaymentTransactionService.java:
```diff
@@ -0,0 +1,56 @@
+            if (!available) {
+                throw new RuntimeException("lock 획득 실패");
+            }
+
```
[중복 request 방지]
- 중복 request 방지를 위해 resson 적용 ( 해당 부분은 AOP 구현해 annotation으로 간편하게 추가하는 방식으로 진행하는게 맞을까요?? )

f-lab-justin replied:
@spongeboom 
넵~ 이후에 커스텀 어노테이션으로 구현해보시지요!

f-lab-justin commented on fpay-api/src/main/java/com/flab/fpay/api/pay/service/PayInfoService.java:
```diff
@@ -0,0 +1,26 @@
+    private final PayInfoRepository payInfoRepository;
+
+    public PayInfo getPayInfoByPaymentTypeInfoId(BigInteger paymentTypeInfoId){
+        return payInfoRepository.findPayInfoByPaymentTypeInfoId(paymentTypeInfoId).orElseThrow(() -> new NoSuchElementException("PayInfo not found with payInfoTypeId: $id"));
```
@spongeboom 
요 라인은 옆으로 너무 기네요.
최대한 좌우 위아래 스크롤 없이 코드를 볼 수 있도록
코딩 스타일 등을 챙겨주세요~

IntelliJ의 code style을 좀 만져보셔도 좋습니다!

f-lab-justin commented on fpay-api/src/main/java/com/flab/fpay/api/pay/PayController.java:
```diff
@@ -24,8 +27,11 @@ public ResponseEntity<PaymentRequestResDTO> paymentReady(@RequestBody PaymentReq
-        return ResponseEntity.ok(Map.of("result","success"));
+    public ResponseEntity<PaymentTransactionResDTO> paymentApprove(@RequestBody PaymentTransactionDTO paymentTransactionDTO){
+
+        PaymentTransactionResDTO paymentTransactionResDTO = paymentTransactionService.approvePayment(paymentTransactionDTO);
```
@spongeboom 
request객체를 그대로 서비스로 넘기면 어떤 장점과 단점이 있을까요?

f-lab-justin commented on fpay-api/src/main/java/com/flab/fpay/api/pay/approve/ApproveMoney.java:
```diff
@@ -0,0 +1,40 @@
+    public PaymentTransaction approve(PaymentTransactionDTO paymentTransactionDTO, PaymentRequest paymentRequest) {
+        PaymentTypeInfo paymentTypeInfo = paymentTypeInfoService.getPaymentTypeInfoByUid(paymentTransactionDTO.getUid());
+        PayInfo payInfo = payInfoService.getPayInfoByPaymentTypeInfoId(paymentTypeInfo.getPaymentTypeInfoId());
+        payInfo.setMoney(payInfo.getMoney().subtract(paymentTransactionDTO.getProductPrice()));
```
@spongeboom 
함수형 코딩을 읽고있는 사람에게 있어서 return값이 없는 함수를 사용하거나, 절차형으로 한라인 한라인 작성하는 것도 어색합니다. 
함수의 합성으로 개발을 하게 되면 하나의 블록과 같이 코드를 짜게 되는 경우가 많습니다.
어떻게 수정해볼 수 있을까요?(map을 사용해볼 수 있을까요?)

f-lab-justin commented on fpay-api/src/main/java/com/flab/fpay/api/pay/service/PaymentTransactionService.java:
```diff
@@ -0,0 +1,56 @@
+
+            if (!paymentTransactionDTO.getCompanyOrderNumber().equals(paymentRequest.getCompanyOrderNumber()) &&
+                    !paymentTransactionDTO.getCompanyUserId().equals(paymentRequest.getCompanyUserId()) &&
+                    !paymentTransactionDTO.getProductPrice().equals(paymentRequest.getPaymentPrice())){
```
@spongeboom 
코드가 좀 장황한데, 좀 더 깔끔하게 작성할 수 있는 방법은 없을까요?

f-lab-justin commented on fpay-api/src/main/java/com/flab/fpay/api/pay/service/PaymentTransactionService.java:
```diff
@@ -0,0 +1,56 @@
+                throw new RuntimeException("결제 요청 정보와 승인 정보가 다릅니다.");
+            }
+
+            if (paymentRequest.getPaymentType().equals("MONEY")){
```
@spongeboom 
Type의 경우에는 Enum을 쓰면 Equals를 사용하지 않아도 되고 실수를 줄일 수 있습니다.

f-lab-justin commented on fpay-api/src/main/java/com/flab/fpay/api/pay/service/PaymentTypeInfoService.java:
```diff
@@ -0,0 +1,22 @@
+    private final PaymentTypeInfoRepository paymentTypeInfoRepository;
+
+    public PaymentTypeInfo getPaymentTypeInfoByUid(BigInteger uid){
+        return paymentTypeInfoRepository.findPaymentTypeInfoByUid(uid).orElseThrow(() -> new NoSuchElementException("PaymentTypeInfo not found with uid: $id"));
```
@spongeboom 
이 라인도 너무 기네요T.T

f-lab-justin commented on fpay-api/src/test/java/com/flab/fpay/api/PayMentDocumentTest.java:
```diff
@@ -132,32 +122,39 @@ public void ready_payment() throws Exception {
+        paymentTypeInfo.setPaymentTypeId(BigInteger.valueOf(1));
         paymentTypeInfo.setUid(BigInteger.valueOf(1));
         paymentTypeInfo.setCreatedAt(LocalDateTime.now());
         paymentTypeInfo.setUpdatedAt(LocalDateTime.now());
```
@spongeboom 
먼가..set함수가 너무 많은데요.
set함수는 최대한 별도로 사용하지 않는게 좋죠? 
그리고 테스트 코드는 명료하게 테스트하고자 하는 로직이 잘 보여지는 것이 좀 더 좋을 것 같습니다.
어떻게 하면 좀 더 간결하면서도 의미 있게 작성할 수 있을까요?

f-lab-justin reviewed (changes requested):
@spongeboom 
한번 생각해보세요~!

IMQA-SUNGYOUN committed:
피드백 반영
- Service
  - code style 적용 ( google java style )
- ApproveMoney
  - Optional map으로 변경

spongeboom commented on fpay-api/src/main/java/com/flab/fpay/api/pay/approve/ApproveMoney.java:
```diff
@@ -0,0 +1,32 @@
+    public PaymentTransaction approve(PaymentTransactionDTO paymentTransactionDTO,
+        PaymentRequest paymentRequest) {
+
+        return Optional.of(
+                paymentTypeInfoService.getPaymentTypeInfoByUid(paymentTransactionDTO.getUid()))
+            .map(paymentTypeInfo -> payInfoService.getPayInfoByPaymentTypeInfoId(
+                paymentTypeInfo.getPaymentTypeInfoId()))
+            .map(payInfoService::savePayInfo)
+            .map(payInfo -> paymentTransactionDTO.toEntity(paymentRequest, payInfo))
+            .orElseThrow(() -> new NoSuchElementException("결제에 실패하였습니다."));
```
피드백 수정( 절차형으로 한라인 한라인 작성 하는 부분 ) 
이 부분 Optional map을 사용하도록 수정 하였습니다.

spongeboom commented on fpay-api/src/main/java/com/flab/fpay/api/pay/service/PayInfoService.java:
```diff
@@ -0,0 +1,28 @@
+
+    private final PayInfoRepository payInfoRepository;
+
+    public PayInfo getPayInfoByPaymentTypeInfoId(BigInteger paymentTypeInfoId) {
+        return payInfoRepository.findPayInfoByPaymentTypeInfoId(paymentTypeInfoId).orElseThrow(
+            () -> new NoSuchElementException("PayInfo not found with payInfoTypeId: $id"));
+    }
```
피드백 수정 ( 코드 라인이 옆으로 너무 긴부분 ) 
- intelij google java code style 적용하여 수정 하였습니다.

f-lab-justin commented:
@spongeboom 
고생하셨습니다! 머징 가시지요!!
